sentry-android-core/src/main/java/io/sentry/android/core/AndroidTransactionProfiler.java

```diff
@@ -229,41 +232,75 @@ public void onFrameMetricCollected(
     scheduledFinish =
         options
             .getExecutorService()
-            .schedule(() -> onTransactionFinish(transaction, true), PROFILING_TIMEOUT_MILLIS);
+            .schedule(
+                () -> timedOutProfilingData = onTransactionFinish(transaction, true),
+                PROFILING_TIMEOUT_MILLIS);
 
     transactionStartNanos = SystemClock.elapsedRealtimeNanos();
     profileStartCpuMillis = Process.getElapsedCpuTime();
 
-    ProfilingTransactionData transactionData =
+    currentProfilingTransactionData =
         new ProfilingTransactionData(transaction, transactionStartNanos, profileStartCpuMillis);
-    transactionMap.put(transaction.getEventId().toString(), transactionData);
 
     Debug.startMethodTracingSampling(traceFile.getPath(), BUFFER_SIZE_BYTES, intervalUs);
   }
 
   @Override
-  public synchronized void onTransactionFinish(final @NotNull ITransaction transaction) {
-    options.getExecutorService().submit(() -> onTransactionFinish(transaction, false));
+  public @Nullable synchronized ProfilingTraceData onTransactionFinish(
+      final @NotNull ITransaction transaction) {
+    try {
+      return options
+          .getExecutorService()
+          .submit(() -> onTransactionFinish(transaction, false))
+          .get();
```

Comment:
hmm, why do we need to use ExecutorService here if we are anyway awaiting for it in a blocking way?

Reply:
to avoid calling `onTransactionStart` in the executor service thread and `onTransactionFinish` in a different thread